### ISSUE TITLE
Preparing for 3.0.0 using 2.8.0?

### ISSUE THREAD

**Issue by SimpleAddress4390** [vubiostat/redcapAPI#246]
Forgot my stupidity up front.

I had written a TON of code to handle the version 2.3.0 extract from REDCap.  With much help from many of you, I was able to get it to correctly cast things.

With 2.8.0, I seem to be back at square one.

The following call works fantasticailly !!!   I save hundreds of lines of code:
 exportBulkRecords(list(db = rcon), envir = 1, cast = raw_cast)

unfortunately, all the fields are coming back as character, even though in 2.3.0, doing the form by form returned them correctly.  (dates as dates, radio buttons, as numeric, multiple choice as numeric, etc.)

I have tried to read the documentation, but can't quite get my head around this.  Is the only way to do this to NOT use exportBulkRecords, and instead find the forms in REDCap, then for each form use ExportRecordsType, and somehow create my own 'casting' function for each variable and say what it should be based on the metadatra?  I'll admit this is a bit beyond me.    I thought the API was already doing some deep logic thingy to extract in the correct type.

Am I just using all these API calls incorrectly?

OLD code that just handled dates, the other stuff came out as numeric when appropriate:

  .... (tons of code)
         dataset <- res
         
         # Get variable names from data dictionary
         tempfieldlist <- dd %>% filter(ExcludeFlag==FALSE & form_name == f[i]) %>% pull(field_name)

         # If the subject ID is not on the dataset, add
         if (tempfieldlist[1] != xsid) {tempfieldlist <- c(xsid  ,tempfieldlist)} 
         
         # Handle date columns where there are date values
         
         dtVect <-dd %>% 
            filter(str_detect(text_validation_type_or_show_slider_number, "date_") & form_name == f[i]) %>% 
            select(field_name) %>% pull()
         
         colclassesList <- replicate(length(dtVect), 
                                "character",
                                simplify = FALSE)
         
         names(colclassesList) <- dtVect
         
         
         # slow it down so we do not get banned from the server
         Sys.sleep(1)
         
         # Extract data again with necessary fields

               dataset <- as.data.frame(exportRecords(
                  rcon,
                  factors = FALSE,
                  fields = tempfieldlist,
                  forms = f[i],
                  records = subjectIDs,
                  events = NULL,
                  labels = FALSE,
                  dates = datesMethod,
                  survey = TRUE,
                  dag = FALSE,
                  checkboxLabels = FALSE,
                  colClasses = colclassesList,
                  batch.size = -1,
                  bundle = getOption("redcap_bundle"),
                  error_handling = getOption("redcap_error_handling"),
                  form_complete_auto = TRUE)
               )    
               
               dataset_wfactors <- as.data.frame(exportRecords(
                  rcon,
                  factors = TRUE,
                  fields = tempfieldlist,
                  forms = f[i],
                  records = subjectIDs,
                  events = NULL,
                  labels = FALSE,
                  dates = datesMethod,
                  survey = TRUE,
                  dag = FALSE,
                  checkboxLabels = FALSE,
                  colClasses = colclassesList,
                  batch.size = -1,
                  bundle = getOption("redcap_bundle"),
                  error_handling = getOption("redcap_error_handling"),
                  form_complete_auto = TRUE)
               )    
     ... (tons more code)

**Comment by nutterb** [comment]
I'm sorry for the frustration this has caused you. You're correct to observe that things are quite a bit different. (2.3.0 was released in 2020, and the change pushed this year have significant improvements).  

The good news is that I think we can get you where you want to be with just a few small changes. 

The crux of your issue is that you are using the argument `cast = raw_cast`. The `raw_cast` list was intended to cast all fields using `as.character`, which seems to be what you are getting. 

If you change your `cast` argument to the following, I think you should get your intended result.

`cast = list(radio = castCode, dropdown = castCode)`. 

This utilize the default casting specification (see https://github.com/vubiostat/redcapAPI/blob/main/R/fieldValidationAndCasting.R#L529-L555) for radio and dropdown while leaving all of the other defaults as they are. 


Now, IF that helps to resolve your question, I'd be interested in knowing what documentation you had been reviewing to try to resolve this yourself.  Specifically, what documentation could we improve to make this easier to find in the future?

**Comment by spgarbet** [comment]
The casting argument is a list of casting functions. The big design difference from the earlier is this moves it away from flags and allows the user full control. 

:100: _Specifically, what documentation could we improve to make this easier to find in the future?_

**Comment by SimpleAddress4390** [comment]
Hi.  First of all, thank you profusely.  I created a new 'list' object and the data extracts as expected (except for the "redcap_repeat_instance" field which comes as character, but I would think should be a number.  I manually corrected via code.  I need to do some in-depth checking, but it visually looks pretty good. 

I need some time to collect my thoughts and maybe write up the docs I looked and (and how I examined them) to explain why I didn't pick up on the cast info.    I do hope to give you feedback (and any help I can). 

regards

```
   special_cast <- list(
      date_                    = function(x, ...) as.POSIXct(x, format = "%Y-%m-%d"),
      datetime_                = function(x, ...) as.POSIXct(x, format = "%Y-%m-%d %H:%M"),
      datetime_seconds_        = function(x, ...) as.POSIXct(x, format = "%Y-%m-%d %H:%M:%S"),
      time_mm_ss               = function(x, ...) chron::times(ifelse(is.na(x),NA,paste0("00:",x)), format=c(times="h:m:s")),
      time_hh_mm_ss            = function(x, ...) chron::times(x, format=c(times="h:m:s")),
      time                     = function(x, ...) chron::times(gsub("(^\\d{2}:\\d{2}$)", "\\1:00", x), 
                                                               format=c(times="h:m:s")),
      float                    = as.numeric,
      number                   = as.numeric,
      number_1dp               = as.numeric, 
      number_1dp_comma_decimal = castDpNumeric(),
      number_2dp               = as.numeric, 
      number_2dp_comma_decimal = castDpNumeric(),
      calc                     = as.numeric,
      int                      = as.integer,
      integer                  = as.numeric,
      yesno                    = castDpNumeric(),
      truefalse                = function(x, ...) x=='1' | tolower(x) =='true',
      checkbox                 = castChecked,
      form_complete            = castLabel,
      select                   = castLabel,
      radio                    = castDpNumeric(),
      dropdown                 = castDpNumeric(),
      sql                      = NA, 
      system                   = castLabel
   )

   exportBulkRecords(list(db = rcon), envir = 1, cast = special_cast)

```

**Comment by nutterb** [comment]
@SimpleAddress4390, have you had a chance to collect any thoughts on our documentation?

**Comment by SimpleAddress4390** [comment]
[redcapAPI-doc info.docx](https://github.com/vubiostat/redcapAPI/files/13188492/redcapAPI-doc.info.docx)

@nutterb  and @spgarbet   Sorry for the delay.  I don't know how helpful this is.  As I say, I if I knew some way to help, I'm all in.

regards

**Comment by SimpleAddress4390** [comment]
Hi.  Have you seen this API?  https://epicentre-msf.github.io/redcap/reference/meta_factors.html   This functionality in the redcapAPI would be quite useful.  Currently, users have to manually parse out the data dictionary to do this.  I've noticed that my code no longer works the same with 2.8.0 and my factor labels are incorrectly aligned with the actual data values.  I'd normally just say, "Well, another day, another R package" but this one is only on github, so I can't see asking users to install and use in case it disappears...

**Comment by nutterb** [comment]
On a separate note, based on your feedback, I started drafting another vignette to give a discussion specific to casting fields in `exportRecordsTyped` and targeted at the new user. Before I get too much further down that path, would you be willing to look it over ([link to pdf](https://github.com/vubiostat/redcapAPI/blob/issue-246-casting-longform-doc/inst/vignette-source/CastingData.pdf)).  I'd be interested in knowing

1. Would such a document have made it easier for you to get started with `exportRecordsTyped`?
2. I believe more discussion of casting would be of use to a newer user.  Should validations and/or missing value detection be given a similar treatment? Or would that be another case of information overload?

Thanks for spending to the time to put together such a detailed and thoughtful response.  I admit that I struggle to imagine what it must be like to experience the new features as user, and truly appreciate hearing your own experience adapting to them.

**Comment by SimpleAddress4390** [comment]
Thank you for all you do!  I will try to look it over as soon as possible
and respond specifically to the 2 questions.  Is it OK if I paste it into
Word and do markups/comments?

regards

**Comment by nutterb** [comment]
> Thank you for all you do! I will try to look it over as soon as possible and respond specifically to the 2 questions. Is it OK if I paste it into Word and do markups/comments? 

Absolutely

**Comment by jwilliman** [comment]
Hello, I hope you don't mind me jumping in.

As someone who is also converting [existing code](https://github.com/jwilliman/redcapTidy) to handle the new `exportRecordsTyped` (although I'm excited to see that your changes will replace much of my previous code!), a greater discussion specific to casting fields is very helpful. I followed the link to your pdf above and found the answer to my question within less than a minute (which was 'How do control the casting of redcap_event_name?')

In short, what you do is fantastic. REDCap is becoming the goto platform for data collection at my University and I'll be assisting many researchers (primarily from a clinical background with little data management expertise) in using your package to extract their data in a usable format. 

1. Yes, a document specific to casting fields would be very useful when starting to use `exportRecordsTyped` (I was searching for it yesterday while getting my head about the recent changes).
2. I'm not sure about validation and missing codes - mostly because these aren't very well used in the projects I've been on. In my experience validation errors normally occur when validation is added retrospectively to a project already in data collect, usually because data collect has begun while the project is still in development! I was pleased with the informative validation warning message I received when exporting data into R.

Thanks very much :)

**Comment by nutterb** [comment]
Thanks @jwilliman and @couthcommander . Based on your comments, I'll prepare separate vignettes for casting, validation, and missing data detection. 

Speaking of which, I revised the [casting vignette](https://github.com/vubiostat/redcapAPI/blob/issue-246-casting-longform-doc/vignettes/precompiled-casting-data.pdf) with better data.  

There's also now a "Getting Started" vignette about [getting connected](https://github.com/vubiostat/redcapAPI/blob/issue-246-casting-longform-doc/vignettes/precompiled-getting-started-connecting.pdf) using `unlockREDCap`.

I'll let these percolate for a week or so before I give them another look and push them.

**Comment by jwilliman** [comment]
Thanks @nutterb 

The vignette is looking great. 

One thought I had was to provide a castLogical option. I realise that this can be written fairly easily by the user as a custom function, but it is one I'd want to apply to checkbox, yesno, and truefalse fields. Currently I'm using: 

    yesno     = function(x, ...) x == '1' | tolower(x) == 'yes',
    truefalse = function(x, ...) x == '1' | tolower(x) == 'true'
    checkbox  = function(x, ...) redcapAPI::castRaw(x, ...) |> as.logical()

I also found myself looking for a list of all the casting option in the appendix of the casting vignette, similar to what is presented in the help page for `fieldValidationAndCasting`. This would quickly allow the user to see what is already available and what might require custom code.

Just my 2 cents! Cheers.

**Comment by spgarbet** [comment]
@SimpleAddress4390 This is exactly the sort of feedback we're looking for. That's a nice piece of work and we appreciate the engagement. We might consider updating the validations in general to include the relaxed set, as it seems like negative values are perfectly fine in any context even the more recent. There are also callbacks for defining what is an NA. If -8 is NA and that's how you want it coded that can be used to set values to NA pre-validation. It however should be default be passing validation if it's an allowed code value.

@nutterb thoughts? should we split this off to another ticket?

**Comment by nutterb** [comment]
I think @SimpleAddress4390 is now the leading expert for customizing validations.  That's some impressive work.

Given the incredible amount of effort, it pains me a little to say that I think there is a substantially easier approach to this, and I think we likely need to make a section for missing values in the documentation of `fieldValidationAndCasting` dedicated to missing value detection. It should also probably highlight that projects can have project-level missing value codes (a feature I'm not sure I've encountered before and, regrettably, are not accessible via the API). 

**Applying project-level missing value codes**

We've recently introduced this scenario in the Frequently Asked Questions of the `vignette("redcapAPI-missing-data-detection")`. For this particular project, I believe the following should produce the desired data without having to relax the validations:

```
customMissingDetection <- function(x, ...){
  is.na(x) | x == "" | x %in% c(-2, -3, -4 -5, -8))
}

exportRecordsTyped(rcon, 
                   cast = special_cast, 
                   na = na_values(customMissingDetection))
```

**Action Items**

Whereas I think this is a failure in our documentation, and I already have some documentation clean up to do to clear up this issue (consolidating the FAQs into a single document). I'm not sure I would recommend adding these specific relaxed validations, as they are somewhat project specific. In particular, REDCap offers codes like "NI", "NIV", "UNK" as recommendations, and these validations would not be able to pick them up. The action items I take away from this are

1. Consolidate FAQs into a single document
2. Add a section to `fieldValidationAndCasting` discussing missing value codes. Mention specifically that project-level missing data codes need to be handled by the user. (Missing data codes are defined in Project Setup >Additional Customizations > Missing Data Codes)
3. Add a section in the Missing Value Detection vignette about project level missing values, with screenshots for where they are defined and displayed in the codebook.
